### ISSUE TITLE
Convert sales order into purchase order

### ISSUE THREAD

**Issue by Robert-Ver** [Dolibarr/dolibarr#24610]
### Feature Request

Hi,
 It would be a great advantage if a sales order could be converted into a purchase order from the relevant sales order. 
 This saves a lot of time and prevents errors when entering.

### Use case

The sales order is copied to a purchase order template. If desired, products can still be added or removed here.

### Suggested implementation

Add a button to the approved sales order for conversion.

**Comment by JonBendtsen** [comment]
seems to already be implemented - except for the "preventing errors" :-D #28962

![image](https://github.com/Dolibarr/dolibarr/assets/2551481/ea0c8f5e-f217-4c7c-9bed-fee7d8fcec85)

@ksar-ksar  can this issue be closed?

**Comment by jman482002** [comment]
> seems to already be implemented - except for the "preventing errors" :-D #28962
> 
> ![image](https://private-user-images.githubusercontent.com/2551481/320065518-ea0c8f5e-f217-4c7c-9bed-fee7d8fcec85.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMwNTg5NDUsIm5iZiI6MTcyMzA1ODY0NSwicGF0aCI6Ii8yNTUxNDgxLzMyMDA2NTUxOC1lYTBjOGY1ZS1mMjE3LTRjN2MtOWJlZC1mZWU3ZDhmY2VjODUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MDdUMTkyNDA1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZmI5MTU5N2NhZDA0ZGZmOGM1Mjg1MTE5YjI1NTkxNDNhN2YyODhjY2FlNTBlNGY0NDk1YzM4MDMyZDAxNWE2MCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.oxq8QUAoVnxNN5qAEolAxshhZYzdmWuLLW_RKk4FM2c)
> 
> @ksar-ksar can this issue be closed?

@JonBendtsen I don't have that drop-down option on v19.02. How to enable?

**Comment by JonBendtsen** [comment]
> I don't have that drop-down option on v19.02. How to enable?

Which status is your order in? Have you enabled the vendor module?

**Comment by jman482002** [comment]
Yes Sales Order and Vendor Modules, both enabled. Only option for Validated Sales Order is to create shipment:
![image](https://github.com/user-attachments/assets/2f5ddac5-bbbc-49c5-9de4-4b6f03748be0)

**Comment by jman482002** [comment]
Enabled Invoices module and got a drop-down. But no Purchase Order option:
![image](https://github.com/user-attachments/assets/6b1641ed-d0f2-40d3-9d04-8ca5656bbaf5)

**Comment by jman482002** [comment]
> > I don't have that drop-down option on v19.02. How to enable?
> 
> Which status is your order in? Have you enabled the vendor module?

Status is Validated

**Comment by JonBendtsen** [comment]
> Enabled Invoices module and got a drop-down. But no Purchase Order option:

Are your user authorized to create a purchase order?

**Comment by jman482002** [comment]
> > Enabled Invoices module and got a drop-down. But no Purchase Order option:
> 
> Are your user authorized to create a purchase order?

Yes

@JonBendtsen I think I may have figured out the descrepancy. Are you in the Purchase Order screen when experiencing this drop-down?

If so, this is not according to the topic: **Convert sales order into purchase order** (from Sales Order)

@ksar-ksar I would suggest re-opening this topic bc it would be a seriously awesome improvement!

**Comment by JonBendtsen** [comment]
https://www.dolibarr.org/forum/t/create-purchase-order-from-sales-order/25320/12

**Comment by JonBendtsen** [comment]
Module list extracted using the API

```
["adherent","agenda","api","banque","barcode","blockedlog","bom","bookmark","categorie","clicktodial","commande","comptabilite","contrat","cron","dav","debugbar","don","ecm","emailcollector","eventorganization","expedition","expensereport","export","externalrss","externalsite","facture","fckeditor","ficheinter","fournisseur","geoipmaxmind","gravatar","holiday","hrm","import","incoterm","knowledgemanagement","ldap","loan","mailing","mailmanspip","margin","modulebuilder","mrp","multicurrency","notification","oauth","opensurvey","paymentbybanktransfer","paypal","prelevement","printing","product","productbatch","projet","propal","receiptprinter","reception","recruitment","resource","salaries","service","socialnetworks","societe","stock","stripe","supplier_proposal","syslog","takepos","tax","ticket","user","variants","webhook","website","workflow","workstation"]
```

**Comment by JonBendtsen** [comment]
FYI: my demo dolibarr is still running 19.0.1 so my pictures is from there

**Comment by sonikf** [comment]
Hi @jman482002 
19.0.2 working instance here, works as expected!
Sorry not a bug!
Have you checked your permissions?

**Comment by jman482002** [comment]
Hello @sonikf 
Are you running a docker image as well?

**Comment by sonikf** [comment]
No! 
My main instance is in a shared host and my secondary( for testing purposes) is a local ubuntu installation.

**Comment by sonikf** [comment]
Just tested in 20.0.0, also works as expected

**Comment by jman482002** [comment]
There must be an explanation out there for how to enable this feature. Anyone?

**Comment by jman482002** [comment]
Finally figured it out!
For the record, **the Sales Order must have at least one line of service!?!?**
This helped:
[https://github.com/Dolibarr/dolibarr/issues/29782](https://github.com/Dolibarr/dolibarr/issues/29782)

What a bizarre choice! (I'll call it a bug bc it makes no sense to me.)

I sell mostly products (physical items). This is why it did not work for me. In order to get it to work, I had to add a mock service item:
![image](https://github.com/user-attachments/assets/79c40ea2-3048-4feb-92bf-faddbbf9af9d)


I looks like this might have been fixed since my version v19.02. Can anyone confirm these findings?

Thanks to everyone.

**Comment by sonikf** [comment]
Hi @jman482002 
You are right, I always have shipping expenses!
I have submitted a pr to backport this to 19
https://github.com/Dolibarr/dolibarr/pull/30597
Please close this.
